rfcs/modal-component.md

```diff
@@ -0,0 +1,154 @@
+* `overrides: {Root, Backdrop, Dialog, DialogInner, Close}` - Optional
+  Overrides for presentational components. See "Presentational Components Props API" below.
+  * `[ComponentName]: ReactComponent | {props: {}, style: {}, component: ReactComponent}` - Optional
+
+### <ModalHeader/> API
+
+TBD
+
+### <ModalBody/> API
+
+TBD
+
+### <ModalFooter/> API
+
+TBD
+
+### SIZE Constant
+
+* `small` - 400px width
+* `medium` - 550px width
+* `large` - 800px width
+* `full` - Span full viewport, minus some margin
+* `cover` - Span full viewport, no margin
+* `auto` - Based on content width
+
+^ Open to ideas here. Too many options?
+
+### Presentational Components Props API
+
+Next properties are passed to every presentational (styled) component
+
+* `$isOpen: boolean`
+* `$size: string|number`
+* `$role: string`
+* `$closable: boolean`
+* `$theme: theme`
+
+More props likely TBA as development continues
+
+### Usage
+
+Basic usage:
+
+```javascript
+import * as React from 'react';
+import {Modal, SIZE} from 'baseui/modal';
+
+class App extends React.Component {
+  state = {
+    open: true
+  };
+  render() {
+    return (
+      <div>
+        <Button onClick={() => this.setState({open: true})}>Open</Button>
+        <Modal
+          isOpen={this.state.open}
+          close={() => this.setState({open: false})}
+        >
+          {({close}) => (
+            <>
+              <ModalTitle>Whatsup!</ModalTitle>
+              <ModalBody>This is a modal</ModalBody>
+              <ModalFooter>
+                <Button type=”tertiary” onClick={close}>Close</Button>
+                <Button type=”primary” onClick={this.submit}>Let’s Go</Button>
+              </ModalFooter>
+            </>
+          )}
+        </Modal>
+      </div>
+
+    );
+  }
+}
+```
+
+### Key Design Decisions
+
+* Modal content will not render server side. This is because portals don't exist server-side, and realistically your page shouldn't be loading with a modal open.
+
+### Accesibility
+
+Good read: https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/dialog.html
+
+* Upon opening, first focusable item in modal will receive focus (form input, footer button, etc)
+* Dialog element will `aria-modal="true"`
+* Explicitly expose props for role, aria-labeledby, aria-describedby, id, to encourage good a11y practices
+* `Tab` key will move between focusable items (form inputs, footer buttons, etc). According to best practices, tabbing past the last focusable item should loop back to the first focusable item.
+* `Escape` key will close the modal
+* Click on backdrop (anywhere outside dialog) will hide modal
+* Background will not be scrollable while modal is open (position: fixed)
+* Focus should ideally be transitioned back to last focused element upon closing
+
+### Possible Future Props
+
+The following props could be useful in the future but are low priority for V1
+
+* `onBackdropClick: function` - Callback handler for when backdrop is clicked specifically
```

Comment:
imo, too many handlers. We may need only two (open\close) and provide reason\source of action in parameters of it

Reply:
If `closeable` is false, then using `onClose` to detect backdrop clicks doesn't make sense. Either way, these extra props here won't be in the initial version, just food for thought for the future.